src/lib/importSystem/download.ts

```diff
@@ -0,0 +1,352 @@
+ *
+ * @param importId - The unique identifier for the import.
+ * @param timeBox - The maximum time allowed for file collection.
+ * @param ftpSettings - The FTP server settings.
+ * @param path - The path on the FTP server to search for files (default is root directory).
+ * @param fileMask - The file mask to filter files (optional).
+ * @returns A Promise that resolves to an array of collected files.
+ * @throws An error if the password is not found for the importId and passwordENV.
+ * @throws An error if failed to create the FTP client.
+ */
+const collectFilesFromSource = async (
+  importId, // The unique identifier for the import
+  timeBox, // The maximum time allowed for file collection
+  ftpSettings, // The FTP server settings
+  path = '/', // The path on the FTP server to search for files (default is root directory)
+  fileMask?: string | undefined, // The file mask to filter files (optional)
+) => {
+  const {
+    host,
+    port,
+    username,
+    password,
+  } = collectServerSettings(importId, ftpSettings);
+
+  let ftpClient;
+  try {
+    ftpClient = new FtpClient({
+      host,
+      port,
+      username,
+      password,
+    }); // Create a new FTP client instance with the FTP server settings
+  } catch (error) {
+    throw new Error(`Failed to create FTP client: ${error.message}`);
+  }
+
+  const priorFile = await getPriorFile(importId); // Get the prior file for the import
+
+  try {
+    await ftpClient.connect();
+  } catch (err) {
+    throw new Error(`Failed to connect to FTP: ${err.message}`);
+  }
+
+  let availableFiles: {
+    fullPath: string,
+    fileInfo: FTPFileInfo,
+    stream?: Promise<Readable>,
+  }[];
+  try {
+    availableFiles = await ftpClient.listFiles(
+      path, // The path on the FTP server to search for files
+      fileMask, // The file mask to filter files
+      priorFile, // The prior file for comparison
+      true, // Include directories in the file list
+    ); // Get the list of available files on the FTP server
+  } catch (err) {
+    throw new Error(`Failed to list files from FTP: ${err.message}`);
+  }
+
+  // only files with a stream populated will work
+  const fetchableAvailableFiles = availableFiles.filter(({ stream }) => stream);
+
+  // Record the available files for the import
+  await recordAvailableFiles(importId, fetchableAvailableFiles);
+
+  const startTime = new Date(); // The start time for file collection
+  const timeLimit = new Date(startTime.getTime() + timeBox); // The time limit for file collection
+
+  const { collectedFiles, hasRemainingFiles } = await collectNextFile(
+    importId, // The unique identifier for the import
+    fetchableAvailableFiles, // The list of available files on the FTP server
+    {
+      start: startTime, // The start time for file collection
+      limit: timeLimit, // The time limit for file collection
+    },
+  ); // Collect the next file within the time limit
+
+  try {
+    await ftpClient.disconnect();
```

Comment:
Do we want something like disconnect in a finally?

Reply:
if the disconnect throws, what more could I do. Calling disconnect again just seems like it might not be a good idea.

Reply:
It was more of a question if an exception was thrown before it gets to the disconnect. Ensure that the connection is always closed.